Bearded.Graphics/MeshBuilders/IndexedTrianglesMeshBuilder.cs

```diff
@@ -1,46 +1,19 @@
 using System;
-using Bearded.Graphics.Rendering;
 using Bearded.Graphics.Vertices;
 using OpenTK.Graphics.OpenGL;
 
 namespace Bearded.Graphics.MeshBuilders
 {
-    public sealed class IndexedTrianglesMeshBuilder<TVertex> : IIndexedTrianglesMeshBuilder<TVertex, ushort>, IDisposable
-        where TVertex : struct, IVertexData
+    public interface IIndexedTrianglesMeshBuilder<TVertex, TIndex> : IIndexedMeshBuilder<TVertex, TIndex>
     {
-        private readonly BufferStream<TVertex> vertices;
-        private readonly BufferStream<ushort> indices;
-
-        public IndexedTrianglesMeshBuilder()
-        {
-            vertices = new BufferStream<TVertex>(new Buffer<TVertex>());
-            indices = new BufferStream<ushort>(new Buffer<ushort>());
-        }
-
-        public void Add(
-            int vertexCount, int indexCount,
-            out Span<TVertex> vertices, out Span<ushort> indices, out ushort indexOffset)
-        {
-            indexOffset = (ushort) this.vertices.Count;
-            vertices = this.vertices.AddRange(vertexCount);
-            indices = this.indices.AddRange(indexCount);
-        }
-
-        public IRenderable ToRenderable()
-        {
-            return Renderable.ForVerticesAndIndices(vertices, indices, PrimitiveType.Triangles);
-        }
-
-        public void Clear()
-        {
-            vertices.Clear();
-            indices.Clear();
-        }
+    }
 
-        public void Dispose()
+    public sealed class IndexedTrianglesMeshBuilder<TVertex>
+        : IndexedMeshBuilder<TVertex>, IIndexedTrianglesMeshBuilder<TVertex, ushort>, IDisposable
+        where TVertex : struct, IVertexData
+    {
+        public IndexedTrianglesMeshBuilder() : base(PrimitiveType.Triangles)
```

Comment:
The primitive type is the only difference, and if this was new code I would not have the subclasses. However, it's public API that other projects (e.g. TD) have a source dependency on, and changing this public API would trickle down a lot.

If it bothers you a lot, I could do that of course.

Reply:
I get your point about the semantics of the interface being rather unclear though. The question is if I should address this here or separately.

Reply:
The more I think about it, the more I have concerns with changing the interface though:

- We're using spans for very efficient writing to the correct memory right now, I'm unsure if that would be possible if we wanted to expose say triplets of values.

- There many be legit use cases for unusual numbers of indices, for example for [primitive restart](https://www.khronos.org/opengl/wiki/Vertex_Rendering#Primitive_Restart). Those could of course be handled more semantically for primitive types where we expect them to be used.

While this second concern is something that seems less critical, the first might be tricky. It may be possible using some marshal/pointer shenanigans, but I'd have to do more research and experiment.

Do you think I should do that to try and improve these interfaces? (At the cost of probably writing some slightly esoteric unsafe code, and possible performance overheads of creating and pushing struct of indices, instead of the indices themselves.)

The only other idea, which would definitely work, would be to just check if the number of indices added divides by 3, for example - but that doesn't seem to add much value to me.

Thoughts?